pico/services/src/main/java/io/helidon/pico/services/DefaultServices.java

```diff
@@ -82,40 +82,46 @@ class DefaultServices implements Services, ServiceBinder, Resettable {
     }
 
     @SuppressWarnings({"unchecked", "rawtypes"})
-    static <T> List<T> explodeAndSort(Collection<?> coll,
-                                      ServiceInfoCriteria criteria,
-                                      boolean expected) {
-        List result;
-
+    static <T> List<T> explodeFilterAndSort(Collection<?> coll,
```

Comment:
this is package private - I can even make this private if it helps. At some point, I need to translate the wildcard types to type T that is being returned - and this is the place I do that.

Reply:
`Services.default <T> ServiceProvider<T> lookupFirst(ServiceInfoCriteria criteria)` is public and part of API - and this is the one that is the source of `T` parameter (among others).
This `Services.lookupFirst` is the reason why you do wrong code, and it should not exist.
Either we need typed `ServiceInfoCriteria`, or it should return `ServiceLProvider<Object>` instead of a typed instance, as we cannot coerce the type correctly into the type used by calling site

Reply:
https://github.com/helidon-io/helidon/issues/6567 logged